modules/widgets/wordpress-post-widget.php

```diff
@@ -27,138 +73,600 @@ public function __construct() {
+		 */
+		if ( ! empty( $original_data ) ) {
+			$widget_data = $original_data;
+		}
+		else {
+			$widget_data = array(
+				'site_info' => array(
+					'last_check'  => null,
+					'last_update' => null,
+					'error'       => null,
+					'data'        => array(),
+				),
+				'posts'     => array(
+					'last_check'  => null,
+					'last_update' => null,
+					'error'       => null,
+					'data'        => array(),
+				)
+			);
+		}
+
+		/**
+		 * Update check time and fetch site information.
+		 */
+		$widget_data['site_info']['last_check'] = time();
+
+		$site_info_raw_data    = $this->fetch_site_info( $site );
+		$site_info_parsed_data = $this->parse_site_info_response( $site_info_raw_data );
+
+
+		/**
+		 * If there is an error with the fetched site info, save the error and update the checked time.
+		 */
+		if ( is_wp_error( $site_info_parsed_data ) ) {
+			$widget_data['site_info']['error'] = $site_info_parsed_data;
+
+			return $widget_data;
+		}
+		/**
+		 * If data is fetched successfully, update the data and set the proper time.
+		 *
+		 * Data is only updated if we have valid results. This is done this way so we can show
+		 * something if external service is down.
+		 *
+		 */
+		else {
+			$widget_data['site_info']['last_update'] = time();
+			$widget_data['site_info']['data']        = $site_info_parsed_data;
+			$widget_data['site_info']['error']       = null;
+		}
+
+
+		/**
+		 * Update check time and fetch posts list.
+		 */
+		$widget_data['posts']['last_check'] = time();
+
+		$site_posts_raw_data    = $this->fetch_posts_for_site( $site_info_parsed_data->ID );
+		$site_posts_parsed_data = $this->parse_posts_response( $site_posts_raw_data );
+
+
+		/**
+		 * If there is an error with the fetched posts, save the error and update the checked time.
+		 */
+		if ( is_wp_error( $site_posts_parsed_data ) ) {
+			$widget_data['posts']['error'] = $site_posts_parsed_data;
+
+			return $widget_data;
+		}
+		/**
+		 * If data is fetched successfully, update the data and set the proper time.
+		 *
+		 * Data is only updated if we have valid results. This is done this way so we can show
+		 * something if external service is down.
+		 *
+		 */
+		else {
+			$widget_data['posts']['last_update'] = time();
+			$widget_data['posts']['data']        = $site_posts_parsed_data;
+			$widget_data['posts']['error']       = null;
+		}
+
+		return $widget_data;
+	}
+
+
+	/**
+	 * Gets blog data from the cache.
+	 *
+	 * @param string $site
+	 *
+	 * @return array|WP_Error
+	 */
+	public function get_blog_data( $site ) {
+		// load from cache, if nothing return an error
+		$site_hash = $this->get_site_hash( $site );
+
+		$cached_data = $this->wp_get_option( $this->widget_options_key_prefix . $site_hash );
+
+		/**
+		 * If the cache is empty, return an empty_cache error.
+		 */
+		if ( false === $cached_data ) {
+			return new WP_Error(
+				'empty_cache',
+				__( 'Information about this blog is being currently retrieved.', 'jetpack' )
```

Comment:
'Information about this blog is currently being retrieved.' reads a little better to me.